src/ext/php7/engine_hooks.c

```diff
@@ -301,20 +294,47 @@ static void _dd_end_span(ddtrace_span_t *span, zval *user_retval) {
 
         zend_throw_exception_internal(NULL);
     }
+
+    return keep_span;
 }
 
-static void _dd_trace_dispatch(ddtrace_dispatch_t *dispatch, zend_function *fbc, zend_execute_data *execute_data) {
+static void _dd_end_span(ddtrace_span_t *span, zval *user_retval) {
+    ddtrace_dispatch_t *dispatch = span->dispatch;
+    dd_trace_stop_span_time(span);
+
+    bool keep_span = true;
+    if (dispatch->options & DDTRACE_DISPATCH_POSTHOOK) {
+        keep_span = _dd_call_sandboxed_tracing_closure(span, user_retval);
+    }
+
+    if (keep_span) {
+        ddtrace_close_span();
+    } else {
+        ddtrace_drop_span();
+    }
+}
+
+static bool _dd_trace_dispatch(ddtrace_dispatch_t *dispatch, zend_function *fbc, zend_execute_data *execute_data) {
+    ddtrace_span_t *span = ddtrace_open_span(EX(call), dispatch);
+
+    if (dispatch->options & DDTRACE_DISPATCH_PREHOOK && _dd_call_sandboxed_tracing_closure(span, NULL) == false) {
```

Comment:
I've been bitten by bitwise operator precedence in the past; do you mind if we group this?
```suggestion
    if ((dispatch->options & DDTRACE_DISPATCH_PREHOOK) && _dd_call_sandboxed_tracing_closure(span, NULL) == false) {
```